### PR TITLE
Flow9 : perceus (attempt 1)

### PR DESCRIPTION
**Pull request by shkel**
it is not finished. but i would like to continue in master.

added dup/drop (fixed not all cases). no real memory management yet.

flag : mem=1
(-> flow9 XXX.flow cpp=1 mem=1)

### CODE REVIEW

File: tools/flow9/backends/bprogram_lower.flow [hunk 1]
```diff
@@ -163,25 +162,54 @@ bexp2BExpDeclaration(declid : string, typars : [string], b : BExp) -> BExpDeclar
...
 			}
 		}
+		BDupValue(value) : var();
+		BDropValue(value) : var();
+		BReuseValue(value) : var();
 	}
 }
 
-bexp2BStatement(decl : string, fnargs : [string], hasTailcall : ref bool, tailcall : bool, b : BExp) -> BStatement {
+downBCastToReturnValue(exp : BExp, bcast : BCast, wrap : (BExp) -> BExp) -> BExp {
```
Comment (alstrup) [inline tools/flow9/backends/bprogram_lower.flow:175]: Add comment about what "wrap" does. It is not clear to me.

File: tools/flow9/backends/link.flow [hunk 2]
```diff
@@ -1,16 +1,22 @@
 import tools/flow9/backends/rename_ids;
 import tools/flow9/backends/bprogram_lower;
 import tools/flow9/driver;
+import tools/flow9/backends/ref_counting;
 
 export {
 	linkBExpProgram(cache : FlowCache, flowpath : string, shorten_ids : bool, spec : BackendSpec) -> BExpProgram;
 	linkBProgram(cache : FlowCache, flowpath : string, shorten_ids : bool, spec : BackendSpec) -> BProgram;
 	programBModules(cache : FlowCache, file : string, acc : List<BModule>, visited : Set<string>) -> Pair<List<BModule>, Set<string>>;
+	memoryManagement = isUrlParameterTrue("mem");
```
Comment (alstrup) [inline tools/flow9/backends/link.flow:10]: I have used "Tracing" for a place to keep flags before.

**Review by alstrup — COMMENTED** [review]
OK to merge and continue on master.

File: tools/flow9/backends/link.flow (hunk 2 above)
Comment (shkel) [reply tools/flow9/backends/link.flow:10]: the flag is mainly used to generate C ++ code in a new format, not to print debug information. the flag will be removed in the future. (it is needed to compare versions of c ++ files.)

**Pull request merged**